crypto/s2n_fips_rules.c

```diff
@@ -0,0 +1,122 @@
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_result.h"
+
+/* FIPS requires at least 112 bits of security.
+ * https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-107r1.pdf */
+const s2n_hash_algorithm fips_hash_algs[] = {
+    S2N_HASH_SHA224,
+    S2N_HASH_SHA256,
+    S2N_HASH_SHA384,
+    S2N_HASH_SHA512,
+};
+S2N_RESULT s2n_fips_validate_hash_algorithm(s2n_hash_algorithm hash_alg, bool *valid)
+{
+    RESULT_ENSURE_REF(valid);
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_hash_algs); i++) {
+        if (fips_hash_algs[i] == hash_alg) {
+            *valid = true;
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+/* https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-52r2.pdf */
+const uint8_t fips_cipher_suite_ianas[][2] = {
+    /* 3.3.1.1.1 Cipher Suites for ECDSA Certificates */
+    { TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 Cipher Suites for RSA Certificates */
+    { TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 3.3.1.2 Cipher Suites for TLS 1.3 */
+    { TLS_AES_128_GCM_SHA256 },
+    { TLS_AES_256_GCM_SHA384 },
+};
+
+S2N_RESULT s2n_fips_validate_cipher_suite(const struct s2n_cipher_suite *cipher_suite, bool *valid)
+{
+    RESULT_ENSURE_REF(cipher_suite);
+    RESULT_ENSURE_REF(valid);
+
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_cipher_suite_ianas); i++) {
+        if (fips_cipher_suite_ianas[i][0] != cipher_suite->iana_value[0]) {
+            continue;
+        }
+        if (fips_cipher_suite_ianas[i][1] != cipher_suite->iana_value[1]) {
+            continue;
+        }
+        *valid = true;
+        return S2N_RESULT_OK;
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_fips_validate_signature_scheme(const struct s2n_signature_scheme *sig_alg, bool *valid)
+{
+    RESULT_ENSURE_REF(sig_alg);
+    RESULT_GUARD(s2n_fips_validate_hash_algorithm(sig_alg->hash_alg, valid));
+    return S2N_RESULT_OK;
+}
+
+/* https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf */
+const struct s2n_ecc_named_curve *fips_curves[] = {
+    &s2n_ecc_curve_secp256r1,
+    &s2n_ecc_curve_secp384r1,
+    &s2n_ecc_curve_secp521r1,
```

Comment:
Strictly speaking there are more curves which are approved here, but I think just keeping these very common prevalent ones is the right thing.

Reply:
Yeah, we're only including options that s2n-tls actually supports. The only other curve we support atm is x25519, which isn't FIPS-compliant.